File: scripts/prediction-sim/lib/streaming-split-planner.mjs
// Pure-logic JS mirror of shared/StreamingSplitPlanner.swift. It tracks only
// sample offsets; the recorder owns audio storage and forced-split replay.

export const STREAM_VAD_MAX_UTTERANCE_SAMPLES = 20 * 16000;
export const STREAM_VAD_USABLE_PAUSE_SAMPLES = 300 * 16;
export const STREAM_VAD_MIN_SPLIT_UTTERANCE_SAMPLES = 1000 * 16;

export class StreamingSplitPlanner {
  constructor({
    maxUtteranceSamples = STREAM_VAD_MAX_UTTERANCE_SAMPLES,
    usablePauseSamples = STREAM_VAD_USABLE_PAUSE_SAMPLES,
    minimumUtteranceSamples = STREAM_VAD_MIN_SPLIT_UTTERANCE_SAMPLES,
    overlapSamples = 0,
  } = {}) {
    this.maxUtteranceSamples = Math.max(0, maxUtteranceSamples);
    this.usablePauseSamples = Math.max(0, usablePauseSamples);
    this.minimumUtteranceSamples = Math.max(0, minimumUtteranceSamples);
    this.overlapSamples = Math.max(0, overlapSamples);
    this.reset();
  }

  begin(initialSamples) {
    this.totalSamples = Math.max(0, initialSamples);
    this.silenceRunSamples = 0;
    this.latestUsablePauseEndSamples = null;
  }

  append(evidence, durationSamples) {
    const samples = Math.max(0, durationSamples);
    if (samples === 0) return;

    if (evidence === 'silence') {
      this.silenceRunSamples += samples;
    } else {
      this.silenceRunSamples = 0;
    }

    const nextTotal = this.totalSamples + samples;
    if (
      evidence === 'silence'
      && this.silenceRunSamples >= this.usablePauseSamples
      && nextTotal >= this.minimumUtteranceSamples
    ) {
      this.latestUsablePauseEndSamples = nextTotal;
    }
    this.totalSamples = nextTotal;
  }

  splitPoint() {
    if (this.totalSamples < this.maxUtteranceSamples || this.maxUtteranceSamples === 0) {
      return null;
    }
    if (this.latestUsablePauseEndSamples !== null) {
      const offset = Math.max(1, this.latestUsablePauseEndSamples - this.overlapSamples);
      return {
        offsetSamples: Math.min(offset, this.maxUtteranceSamples),
        reason: 'pause',
      };
    }
    return { offsetSamples: this.maxUtteranceSamples, reason: 'cap' };
  }

  reset() {
    this.totalSamples = 0;
    this.silenceRunSamples = 0;
    this.latestUsablePauseEndSamples = null;
  }
}

File: shared/StreamingSplitPlanner.swift
import Foundation

enum StreamingSplitReason: String {
    case pause
    case cap
}

struct StreamingSplitPoint {
    let offsetSamples: Int
    let reason: StreamingSplitReason
}

/// Sample-count-only forced-split planning. Audio storage remains owned by the
/// recorder; this type only remembers the latest usable confirmed silence run.
final class StreamingSplitPlanner {
    let maxUtteranceSamples: Int
    let usablePauseSamples: Int
    let minimumUtteranceSamples: Int
    let overlapSamples: Int

    private(set) var totalSamples = 0
    private var silenceRunSamples = 0
    private(set) var latestUsablePauseEndSamples: Int?

    init(
        maxUtteranceSamples: Int,
        usablePauseSamples: Int = 4_800,
        minimumUtteranceSamples: Int = 16_000,
        overlapSamples: Int = 0
    ) {
        self.maxUtteranceSamples = max(0, maxUtteranceSamples)
        self.usablePauseSamples = max(0, usablePauseSamples)
        self.minimumUtteranceSamples = max(0, minimumUtteranceSamples)
        self.overlapSamples = max(0, overlapSamples)
    }

    func begin(initialSamples: Int) {
        totalSamples = max(0, initialSamples)
        silenceRunSamples = 0
        latestUsablePauseEndSamples = nil
    }

    func append(evidence: StreamingEndpointEvidence, durationSamples: Int) {
        let samples = max(0, durationSamples)
        guard samples > 0 else { return }

        if evidence == .silence {
            silenceRunSamples += samples
        } else {
            silenceRunSamples = 0
        }

        let nextTotal = totalSamples + samples
        if evidence == .silence,
           silenceRunSamples >= usablePauseSamples,
           nextTotal >= minimumUtteranceSamples {
            latestUsablePauseEndSamples = nextTotal
        }
        totalSamples = nextTotal
    }

    func splitPoint() -> StreamingSplitPoint? {
        guard totalSamples >= maxUtteranceSamples, maxUtteranceSamples > 0 else {
            return nil
        }
        if let pauseEnd = latestUsablePauseEndSamples {
            let offset = max(1, pauseEnd - overlapSamples)
            return StreamingSplitPoint(
                offsetSamples: min(offset, maxUtteranceSamples),
                reason: .pause
            )
        }
        return StreamingSplitPoint(offsetSamples: maxUtteranceSamples, reason: .cap)
    }

    func reset() {
        totalSamples = 0
        silenceRunSamples = 0
        latestUsablePauseEndSamples = nil
    }
}
